File: src/AI/ImageTagAnalyze.js
// ImageTagAnalyze.js

const { ComputerVisionClient } = require("@azure/cognitiveservices-computervision");
const { CognitiveServicesCredentials } = require("@azure/ms-rest-azure-js");
const { Translate } = require('@google-cloud/translate').v2; // Google Cloud Translation API
const dotenv = require('dotenv');
dotenv.config();

// Google Cloud Translation API 인증 정보 설정
const translate = new Translate({
  projectId: process.env.GCP_PROJECT_ID,
  keyFilename: process.env.TRANSLATE_KEYFILE
});

async function tagAndTranslateImage(imageData) {
    // Azure Cognitive Services 인증 정보 설정
    const subscriptionKey = process.env.VISION_KEY;
    const endpoint = process.env.VISION_ENDPOINT;
    const credentials = new CognitiveServicesCredentials(subscriptionKey);
    const computervisionClient = new ComputerVisionClient(credentials, endpoint);

    try {
        // 이미지 태그 분석
        const tagsResult = await computervisionClient.tagImageInStream(imageData);
        const imageTags = tagsResult.tags.map(tag => tag.name);

        if (imageTags.length === 0) {
            console.log("No tags detected.");
            return [];
        }

        const tagList = [
            "beach",
            "food",
            "person",
            "flower",
            "mountain",
            "sky",
            "night",
            "building",
            "animal",
            "nature",
            "skyscraper",
            "art",
            "concert",
            "tree",
            "plant",
            "river",
            "cloud",
            "car",
            "sun"
        ];

        const enTagResult = imageTags.filter(tag => tagList.includes(tag));

        // 번역
        const koTagResult = await Promise.all(enTagResult.map(async (tag) => {
            try {
                const [translation] = await translate.translate(tag, 'ko');
                return translation;
            } catch (error) {
                console.error("Translation error:", error);
                return tag; // 에러 발생 시 원문 그대로 반환
            }
        }));

        return koTagResult;
    } catch (error) {
        console.error("Error:", error);
        return [];
    }
}

module.exports = {
    tagAndTranslateImage: tagAndTranslateImage
};

File: src/AI/ImageTagAnalyze.py
from azure.cognitiveservices.vision.computervision import ComputerVisionClient
from azure.cognitiveservices.vision.computervision.models import VisualFeatureTypes
from msrest.authentication import CognitiveServicesCredentials
import os
from dotenv import load_dotenv
import googletrans


load_dotenv()


#------------------------------------------------------------------------------------------------------

# Authenticate
# key, endpoint
subscription_key = os.environ.get('VISION_KEY')
endpoint = os.environ.get('VISION_ENDPOINT')
computervision_client = ComputerVisionClient(endpoint, CognitiveServicesCredentials(subscription_key))

# Quickstart variables
images_folder = os.path.join (os.path.dirname(os.path.abspath(__file__)), "images")
remote_image_url = "https://gscaltexmediahub.com/wp-content/uploads/2023/05/the-day-of-ocean-2023_1.png"

# tagging the image
imageTag = []
print("===== Tag an image =====")
tags_result_remote = computervision_client.tag_image(remote_image_url )

print("Tags in the remote image: ")
if (len(tags_result_remote.tags) == 0):
    print("No tags detected.")
else:
    for tag in tags_result_remote.tags:
        print("'{}' with confidence {:.2f}%".format(tag.name, tag.confidence * 100))
        imageTag.append(tag.name)


#------------------------------------------------------------------------------------------------------

tag_list = [
    "beach",
    "food",
    "person",
    "flower",
    "mountain",
    "sky",
    "night",
    "building",
    "animal",
    "nature",
    "skyscraper",
    "art",
    "concert",
    "tree",
    "plant",
    "river",
    "cloud",
    "car",
    "sun"
]

if len(imageTag) % 2 == 0:
    len_ = int(len(imageTag) / 2)
else:
    len_ = int(len(imageTag) / 2) + 1

en_tag_result = []
for i in range(len(imageTag[:len_])):
    for j in range(len(tag_list)):
        if imageTag[i] == tag_list[j]:
            en_tag_result.append(tag_list[j])
print(en_tag_result)

translator = googletrans.Translator()

ko_tag_result = []
for i in en_tag_result:
    result = translator.translate(i, dest="ko")
    ko_tag_result.append(result.text)
print(ko_tag_result)